### DISCUSSION TITLE
Make light grays darker

### DISCUSSION THREAD

**Discussion by DavidTylerPozz** [ImageMagick/ImageMagick#5280] [Help]
Hello - I've converted an image to greyscale and some of the image is a very faint gray against the white bg. I'd like to raise any really light gray to a darker gray but without the lines going jagged (which always seems to happen when I use threshold or monochrome)

I'm using Windows but I don't know what version of IM I've got; happy to add if someone shows me how to find it!

Here's an example of the type of pics I'm trying to selectively darken - see, for example, the very faint arrow marked TIq, it would be great if that could be as dark as, say, the one marked TIp. Or, for example, the box drawn round the equations just by the circled "4"

![result30](https://user-images.githubusercontent.com/108232902/175825369-e1bc5036-8f85-44f4-84a4-83b81afbaca0.png)

**Comment by snibgo** [comment]
"magick -version" will say what version you have. In general, it is best to copy and paste the text output to the forum.

```
magick tensorFld.png -evaluate Pow 3.0 x.png
```

This lowers middle values, without changing black or white. Instead of "3.0", use larger numbers for a bigger difference, or lower numbers down to 1.0 for a lesser effect. Numbers between 0.0 and 1.0 will *increase* middle values, making them lighter.

**Comment by DavidTylerPozz** [comment]
Oh bingo, that worked right out of the box! Fantastic, thank you!

Interestingly the lightest bits do now have that slight jaggedy quality, but the other stuff doesn't. (See the TIq arrow on the "fixed" version at Pow 3.0). So this is still a great solution.

![result32](https://user-images.githubusercontent.com/108232902/175826783-7711a77b-76c5-45eb-a304-3e0648602b06.png)

For completeness, my version is:


Version: ImageMagick 7.1.0-36 Q16-HDRI x64 be0bdc3:20220530 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2022 (193231329)

**Comment by fmw42** [comment]
Invert so black is white and white is black. Threshold at 0. Then invert back. Then since you have an alpha channel, flatten it against white background.

`magick 175825369-e1bc5036-8f85-44f4-84a4-83b81afbaca0.png -negate -threshold 0 -negate -background white -flatten x.png`

![x](https://user-images.githubusercontent.com/13458218/175828067-d1d794bf-5839-486e-b5c6-0ad62d4727ae.png)

On Windows, it might be magick.exe ...

**Reply by fmw42** [reply]
Alternate, since there is data in the alpha channel as well as the underlying image, extract the alpha channel, negate (invert black/white), then threshold at 0.

`magick 175825369-e1bc5036-8f85-44f4-84a4-83b81afbaca0.png -alpha extract -negate -threshold 0 x.png`

![x](https://user-images.githubusercontent.com/13458218/175828202-92194693-7908-4192-9e10-c7ab81430688.png)

**Reply by DavidTylerPozz** [reply]
Ah, thank you so much - I tried this one and it came out a bit jagged and "steppy" - which I think was always the issue whenever I tried any threshold option. It sort of does the job but makes it look cludgy. I guess (I'm no designer) that threshold eats up all that clever shading that lines have as they cross from the grey to the bg and just makes it a bit binary

**Reply by fmw42** [reply]
Correct.  Anything you do globally would make the anti-aliasing go away as it would make that part darker too. If you can isolate just the gray level you need to could make that darker.

So analyzing your image under the transparency, I see two different non-black gray levels. So I extract the alpha channel, then change those two gray levels to black with 5% tolerance in the gray levels. Then I put the alpha channel back and flatten against white.

Windows syntax:

```
magick 175825369-e1bc5036-8f85-44f4-84a4-83b81afbaca0.png  +write mpr:img +delete ^
( mpr:img -alpha extract ) ^
( mpr:img -alpha off -fuzz 5% -fill black -opaque "gray(62.35%)" -opaque "gray(81.57%)" ) ^
+swap -compose copy_opacity -composite ^
-compose over -background white -flatten x.png

```

![x](https://user-images.githubusercontent.com/13458218/175829279-762581bf-5aa0-4e24-a6e8-14b30f30978d.png)

Does that look any better?

**Reply by DavidTylerPozz** [reply]
Ah, anti-aliasing, that's the phrase I was looking for! This is v clever, I tried it and oddly, now _everything_ comes out sort of jittery and step-py (although it looks smooth on your one). 

![result50](https://user-images.githubusercontent.com/108232902/175829522-37bebc7a-6a6f-4336-8046-e5d09bad04ee.png)

**Reply by fmw42** [reply]
Any anti-aliased pixels that have within 5% of those two colors will also change. You can try to reduce the 5% and see if it looks any better.

**Reply by DavidTylerPozz** [reply]
Still got the jitters sadly, even with the fuzz down to virtually nothing. But I'm learning a lot about everything here today, thank you!

**Reply by fmw42** [reply]
We can add some anti-aliasing to the alpha channel by doing some small amount of blur (and stretching the dark values to black).

```
magick 175825369-e1bc5036-8f85-44f4-84a4-83b81afbaca0.png  +write mpr:img +delete ^
( mpr:img -alpha extract -blur 0x1.5 -level 50x100% ) ^
( mpr:img -alpha off -fuzz 5% -fill black -opaque "gray(62.35%)" -opaque "gray(81.57%)" ) ^
+swap -compose copy_opacity -composite ^
-compose over -background white -flatten x.png
```

![x](https://user-images.githubusercontent.com/13458218/175831227-e14bee6a-b14d-4297-88bd-82c835072770.png)

Adjust the blur to 0x1 for less anti-aliasing and 0x2 for more anti-aliasing.

**Comment by DavidTylerPozz** [comment]
Heh, the Blur at 0x1 gives the best combo (the lines tend to be too splindly with 1.5 and above) and the jitter is now very small. Great stuff
![result80lessaa](https://user-images.githubusercontent.com/108232902/175832116-e5257453-abe0-4419-8a1c-2953fea61e8a.png)
